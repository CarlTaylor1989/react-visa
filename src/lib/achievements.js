import _ from 'lodash';
import achievements, { groups } from '../config/achievements';

export const lib = {};

/**
 * Gets achievements sorted by score descending.
 * @return {array}
 */
lib.sortedAchievements = () => _.orderBy(achievements, 'score', 'desc');

/**
 * Gets the percentage complete for an achievement based
 * on how many tasks have been completed.
 * @param {number} completed Total completed
 * @param {number} total     Total tasks
 * @return {number}
 */
lib.getPercentageComplete = (completed, total) => {
  let percent;
  if (completed > 0) {
    const rounded = _.round((completed / total) * 100);
    percent = _.clamp(rounded, 0, 100);
  } else {
    percent = 0;
  }
  return percent;
};

/**
 * Returns achievements with their completion data.
 * @param {object} data The data required to get the achievement status
 * @return {object}
 */
export const getUserAchievementData = (data) => {
  let sortedAchv = lib.sortedAchievements();
  sortedAchv = _.filter(sortedAchv, ach => ach.available !== false);
  return sortedAchv.map((achievement) => {
    const complete = data.completed.includes(achievement.id);
    let completion = {
      complete,
      completedAmount: 0,
      completedPercent: complete ? 100 : 0,
      totalAmount: achievement.max || 1
    };
    switch (achievement.group) {
      case groups.scored: {
        const completedAmount = (data.total > completion.totalAmount)
          ? completion.totalAmount
          : data.total;
        completion = {
          ...completion,
          completedAmount,
          completedPercent: lib.getPercentageComplete(completedAmount, completion.totalAmount)
        };
        break;
      }
      case groups.challenges: {
        if (achievement.max) {
          if (achievement.perfect) {
            completion.completedAmount = complete ? achievement.max : data.noOfPerfectChallenges;
            completion.completedPercent = lib.getPercentageComplete(
              data.noOfPerfectChallenges,
              achievement.max
            );
          } else {
            completion.completedAmount = complete ? achievement.max : data.noOfCompletedChallenges;
            completion.completedPercent = lib.getPercentageComplete(
              data.noOfCompletedChallenges,
              achievement.max
            );
          }
        } else if (achievement.progress && complete) {
          completion.completedAmount = 1;
        }
        break;
      }
      case groups.products: {
        completion = {
          ...completion,
          completedAmount: complete ? achievement.max : data.noOfCompletedProducts,
          completedPercent: lib.getPercentageComplete(data.noOfCompletedProducts, achievement.max)
        };
        break;
      }
      case groups.timed: {
        completion.completedAmount = complete ? achievement.max : data.noOfConsecutiveDays;
        if (!complete) {
          completion.completedPercent = lib.getPercentageComplete(
            data.noOfConsecutiveDays,
            achievement.max
          );
        }
        break;
      }
      case groups.settings: {
        if (completion.complete) {
          completion = {
            ...completion,
            completedAmount: 1
          };
        }
        break;
      }
      case groups.proposals:
        if (achievement.level || achievement.allClients) {
          const completed = data.clients.filter(cl => (
            cl.completed
            && (achievement.allClients || (achievement.level && cl.level === achievement.level))
          ));
          completion.completedAmount = completed.length;
          completion.completedPercent = lib.getPercentageComplete(
            completion.completedAmount,
            achievement.max
          );
        } else {
          completion.completedAmount = completion.complete ? 1 : 0;
        }
        break;
      default: {
        completion.completedAmount = completion.complete ? 1 : 0;
        break;
      }
    }

    return {
      ...achievement,
      completion
    };
  });
};
