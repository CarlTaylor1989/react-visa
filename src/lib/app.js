/**
 * Removes loader a while after component has finished mounting and rendering
 */
import { REMOVE_LOADER_AFTER } from '../config/constants';

export const lib = {};
lib.removeElement = (el) => {
  if (el.parentNode && el.parentNode.removeChild) {
    el.parentNode.removeChild(el);
  }
};

lib.toggleHiddenClass = (el, show) => {
  if (show) {
    el.classList.add('hidden');
  } else {
    el.classList.remove('hidden');
  }
};

/**
 * Hides the loading gif
 */
export const hideLoader = () => {
  const loader = document.getElementById('loader');
  if (loader) {
    setTimeout(() => {
      lib.toggleHiddenClass(loader, true);
    }, REMOVE_LOADER_AFTER);
  }
};

/**
 * Displays the loading gif
 */
export const displayLoader = () => {
  const loader = document.getElementById('loader');
  if (loader) {
    lib.toggleHiddenClass(loader, false);
  }
};

/**
 * Close game SCO window
 */
lib.findAndCloseWindow = (win) => {
  win.close();
  if (win.parent) {
    win.parent.open('', '_self', '');
    win.parent.close();
  }
};

export const closeWindow = () => {
  lib.findAndCloseWindow(window);
};

lib.detectDevice = () => {
  const agent = window.navigator.userAgent;
  const deviceWidth = window.innerWidth
    || document.documentElement.clientWidth
    || document.getElementsByTagName('body')[0].clientWidth;

  // iPhone
  let isIPhone = agent.match(/iPhone/i) != null;

  // iPad up to iOS 12
  /**
   * iPad Pro when run with no launch screen can have error in userAgent reporting as an iPhone
   * rather than an iPad. iPad Pro width portrait 768, iPhone6 plus 414x736 but would probably
   * always report 414 on app startup
   */
  let isIPad = (agent.match(/iPad/i) !== null)
    || ((agent.match(/iPhone/i) !== null) && (deviceWidth > 750));

  if (isIPad) isIPhone = false;

  // iPad from iOS 13
  const isMacSafari = agent.match(/Macintosh/i) != null;
  if (isMacSafari) { // need to distinguish between Macbook and iPad
    const canvas = document.createElement('canvas');
    if (canvas) {
      const context = canvas.getContext('webgl') || canvas.getContext('experimental-webgl');
      if (context) {
        const info = context.getExtension('WEBGL_debug_renderer_info');
        if (info) {
          const renderer = context.getParameter(info.UNMASKED_RENDERER_WEBGL);
          isIPad = renderer.includes('Apple');
        }
      }
    }
  }

  return {
    isIOS: isIPhone || isIPad,
    isIPhone,
    isIPad,
    isMacSafari
  };
};

/**
 * Returns the unload event name based on the device type
 * @return {string}
 */
export const getUnloadEventName = () => {
  let eventName;
  if (lib.detectDevice().isIOS) {
    eventName = 'pagehide'; // before unload is not not supported on iOS
  } else {
    eventName = 'beforeunload';
  }
  return eventName;
};

/**
 * Returns the width and height of the window based on the device
 * @return {object}
 */
export const getScreenSize = () => {
  let width;
  let height;

  if (lib.detectDevice().isIPad) {
    const vw = window.screen.width;
    const vh = window.screen.height;

    // Ensure the viewport width and height is always reported correctly
    if (Math.abs(window.orientation) / 90 === 1) { // Landscape
      width = vw > vh ? vw : vh;
      height = vw > vh ? vh : vw;
    } else { // Portrait
      width = vw > vh ? vh : vw;
      height = vw > vh ? vw : vh;
    }
  } else {
    width = window.innerWidth;
    height = window.innerHeight;
  }
  return { width, height };
};
