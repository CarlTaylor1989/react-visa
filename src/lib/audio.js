export const getAudioPlayer = () => document.getElementById('gameAudioPlayer');

export const getAppRoot = () => {
  const href = window.location.href;
  const appRoot = href.substring(0, href.lastIndexOf('#'));
  return appRoot.substring(0, appRoot.lastIndexOf('/') + 1);
};
