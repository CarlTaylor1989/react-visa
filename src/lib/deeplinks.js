/**
 * Gets the SCO url from a cornestone deeplink
 */
import { LMS_URL } from '../config/constants';

export const lib = {};

const sendRequest = (url, success, error) => {
  const xhr = new XMLHttpRequest();
  xhr.onreadystatechange = () => {
    if (xhr.readyState === 4) {
      if (xhr.status === 200) {
        if (success) {
          success(xhr.responseText);
        }
      } else if (error) {
        error();
      }
    }
  };
  xhr.open('GET', url);
  xhr.send(null);
};
lib.sendRequest = sendRequest;

const getStringBetween = (text, str1, str2) => {
  let str = '';
  const firstPart = text.split(str1);
  if (firstPart.length === 2) {
    const secondPart = firstPart[1].split(str2);
    if (secondPart.length === 2) {
      str = secondPart[0];
    }
  }
  return str;
};
lib.getStringBetween = getStringBetween;

const getFinalLink = (response, callback) => {
  const url = lib.getStringBetween(
    response,
    'coursePopupWindow = popupTitle("',
    '",1'
  );
  callback(url);
};
lib.getFinalLink = getFinalLink;

const getSecondLink = (response, options) => {
  const location = lib.getStringBetween(response, 'url=', '"><title>');
  if (location) {
    const newUrl = LMS_URL + location;
    lib.sendRequest(
      newUrl,
      (responseText) => {
        lib.getFinalLink(responseText, options.success);
      },
      options.error
    );
  } else {
    const url = lib.getStringBetween(
      response,
      'coursePopupWindow = popupTitle("',
      '",1'
    );
    if (url) {
      options.success(url);
    } else {
      options.error();
    }
  }
};
lib.getSecondLink = getSecondLink;

const getFirstLink = (response, options) => {
  const location = lib.getStringBetween(
    response,
    'window.location = \'../scorm/',
    '\';var htmlElem'
  );
  if (location) {
    const newUrl = LMS_URL + location;
    lib.sendRequest(
      newUrl,
      (responseText) => {
        lib.getSecondLink(responseText, options);
      },
      options.error
    );
  } else {
    options.error();
  }
};
lib.getFirstLink = getFirstLink;

export const getDeeplinkURL = (url, options) => {
  lib.sendRequest(
    url,
    (response) => {
      lib.getFirstLink(response, options);
    },
    options.error
  );
};
