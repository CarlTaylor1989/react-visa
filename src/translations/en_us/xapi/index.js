import course from './course.json';
import diagnostic from './diagnostic.json';
import introseq from './introseq.json';
import bonusstreak from './bonusstreak.json';
import ranks from './ranks.json';
import settings from './settings.json';
import map from './map.json';
import products from './products.json';
import dashboard from './dashboard.json';
import gomo from './gomo';
import clientproposals from './clientproposals.json';
import client from './client.json';
import tutorials from './tutorials.json';
import feedback from './feedback.json';

export default {
  course,
  diagnostic,
  introseq,
  bonusstreak,
  ranks,
  settings,
  map,
  products,
  dashboard,
  gomo,
  clientproposals,
  client,
  tutorials,
  feedback
};
