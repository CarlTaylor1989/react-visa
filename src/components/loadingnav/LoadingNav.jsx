import PropTypes from 'prop-types';
import _ from 'lodash';
import { GoToScreen, newScreenConnect } from '../generic/GoToScreen';
import { hideLoader } from '../../lib/app';
import { initialiseLrs } from '../../tracking/xapi/XApiAdapter';
import { indexPath } from '../../config/navigation';
import { gameReferrer } from '../../config/referrers';

export class LoadingNav extends GoToScreen {
  componentDidMount() {
    const { scormInitialised } = this.props;
    if (process.env.PLATFORM === 'local') {
      this.findAndGoToScreen();
    } else if (scormInitialised) {
      this.onScormReady();
    }
  }

  componentDidUpdate(prevProps) {
    const {
      dataReady,
      scormInitialised,
      setScreenReferrer,
      updateTimedAchievements
    } = this.props;

    if (!prevProps.scormInitialised && scormInitialised) {
      this.onScormReady();
    }

    if (prevProps.dataReady !== dataReady) {
      updateTimedAchievements();
      hideLoader();
      setScreenReferrer(gameReferrer);
      this.goToSplashScreen();
    }
  }

  onScormReady() {
    const {
      learnerID,
      learnerName,
      lessonLocation,
      suspendData,
      suspendDataParsed
    } = this.props;

    // Initialise LRS
    if (learnerID && learnerName) {
      initialiseLrs(
        {
          id: learnerID,
          name: learnerName
        },
        lessonLocation !== indexPath
      );
    }

    // Retrieve game setting from the suspend data
    if (suspendData && !_.isEmpty(suspendData)) {
      this.retrieveDataFromSuspendData(suspendData);
    }

    suspendDataParsed(true);
  }

  retrieveDataFromSuspendData(suspendData) {
    const {
      addPrompts,
      restoreAchievements,
      setCategories,
      setDiagnosticData,
      setGenericData,
      setRank,
      setSettings,
      setScores,
      setStreak,
      setClientProposalData,
      setTutorialLaunched
    } = this.props;

    // Set retrieved settings
    if (suspendData.settings) {
      setSettings(suspendData.settings);
    }

    // Set diagnostic completion
    if (suspendData.diagnostic) {
      setDiagnosticData(suspendData.diagnostic);
    }

    // Set retrieved scores
    if (suspendData.score) {
      setRank(suspendData.score.rank);
      setScores(suspendData.score.running, suspendData.score.total);

      if (suspendData.score.categories) {
        setCategories(suspendData.score.categories);
      }
    }

    // Set retrieved streak data
    if (suspendData.streak) {
      setStreak(suspendData.streak);
    }

    // Set retrieved generic data
    if (suspendData.generic) {
      setGenericData(suspendData.generic);
    }

    // Set the retrieved prompt data
    if (suspendData.prompts) {
      addPrompts(suspendData.prompts);
    }

    // set the achievements data
    if (suspendData.achievements) {
      restoreAchievements(
        suspendData.achievements.completed,
        suspendData.achievements.lastCompleted
      );
    }

    // set the client data
    if (suspendData.clientProposalData) {
      setClientProposalData(suspendData.clientProposalData);
    }

    // set the tutorial data
    if (suspendData.tutorial) {
      setTutorialLaunched(suspendData.tutorial.launched);
    }
  }

  render() {
    return null;
  }
}

LoadingNav.propTypes = {
  addPrompts: PropTypes.func.isRequired,
  scormInitialised: PropTypes.bool.isRequired,
  dataReady: PropTypes.bool.isRequired,
  learnerID: PropTypes.string.isRequired,
  learnerName: PropTypes.string.isRequired,
  lessonLocation: PropTypes.string.isRequired,
  restoreAchievements: PropTypes.func.isRequired,
  suspendData: PropTypes.object.isRequired,
  setCategories: PropTypes.func.isRequired,
  setDiagnosticData: PropTypes.func.isRequired,
  setGenericData: PropTypes.func.isRequired,
  setRank: PropTypes.func.isRequired,
  setScreenReferrer: PropTypes.func.isRequired,
  setSettings: PropTypes.func.isRequired,
  setScores: PropTypes.func.isRequired,
  setStreak: PropTypes.func.isRequired,
  suspendDataParsed: PropTypes.func.isRequired,
  updateTimedAchievements: PropTypes.func.isRequired,
  setClientProposalData: PropTypes.func.isRequired,
  setTutorialLaunched: PropTypes.func.isRequired
};

export const mapStateToProps = state => ({
  dataReady: state.genericData.dataReady,
  learnerID: state.scormData.learnerID,
  learnerName: state.scormData.learnerName,
  lessonLocation: state.scormData.lessonLocation,
  suspendData: state.scormData.suspendData,
  scormInitialised: state.scormData.scormInitialised,
});

export default newScreenConnect(LoadingNav, mapStateToProps);
