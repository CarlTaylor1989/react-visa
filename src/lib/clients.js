import { I18n } from 'react-redux-i18n';
import clients, {
  CLIENT_LOCKED,
  STEP1_SCORES,
  STEP2_SCORES,
  REQUIREMENT_MAX_ATTEMPTS
} from '../config/clients';
import { CLIENT_COMPLETED, CLIENT_GIVENUP } from '../config/constants';
import { clientsPropsReferrer } from '../config/referrers';
import { sendInteractionData } from '../tracking/xapi/XApiAdapter';

export const lib = {};

/**
 * Send an xAPI statement for the client proposal product selection
 * @param {string} clientId The client ID
 * @param {array} availableChoices The available choices
 * @param {array} correctResponses The client proposal correct responses
 * @param {array} answers The learner's responses
 * @param {boolean} isCorrect Question answered correctly
 * @param {object} timeAvailable The time the popup became available
 * @param {number} questionIndex The question number
 */
export const sendClientQuestionStatement = (
  clientId,
  availableChoices,
  correctResponses,
  answers,
  isCorrect,
  timeAvailable,
  questionIndex
) => {
  const choices = [];

  availableChoices.forEach((choice) => {
    choices.push({
      id: choice.id,
      description: choice.text
    });
  });
  const problemType = I18n.t(`clientprops.clients.${clientId}.problemType`);

  sendInteractionData({
    type: 'answered',
    activity: {
      id: `${clientsPropsReferrer}/${clientId}/question-${questionIndex}`,
      type: 'choice',
      title: I18n.t('xapi.client.questionTitle', {
        name: I18n.t(`clientprops.clients.${clientId}.info.account`),
        index: questionIndex
      }),
      description: I18n.t('xapi.client.questionDescription', {
        name: I18n.t(`clientprops.clients.${clientId}.info.account`),
        problem: I18n.t(`clientprops.problems.${problemType}.name`),
        question: I18n.t(`xapi.client.question${questionIndex}`)
      }),
      correctResponses: [correctResponses.join(',')],
      choices
    },
    result: {
      completion: isCorrect,
      success: isCorrect
    },
    response: {
      detail: answers.join(','),
      timeAvailable,
      timeResponse: new Date()
    },
    parent: `${clientsPropsReferrer}/${clientId}`
  });
};

/**
 * Calculate number of wins
 * @param {object} data Client proposal state
 * @return {number}
 */
export const calculateTotalWins = (data) => {
  const keys = Object.keys(data);
  return keys
    .map(key => data[key])
    .filter(key => key.proposalCompleted && key.status === CLIENT_COMPLETED).length;
};

/**
 * Calculate number of losses
 * @param {object} data Client proposal state
 * @return {number}
 */
export const calculateTotalLosses = (data) => {
  const keys = Object.keys(data);
  return keys
    .map(key => data[key])
    .filter(key => key.proposalCompleted && key.status === CLIENT_GIVENUP).length;
};

/**
 * Checks if the selected statement is correct
 * @param {string} client The client ID
 * @param {string} statement The statement ID
 * @return {boolean}
 */
export const checkStatementCorrectness = (client, statement) => (
  clients[client].statements.includes(statement)
);

/**
 * Returns the number of correct statements for a given client
 * @param {string} client The client ID
 * @return {number}
 */
export const getNumberOfCorrectStatements = client => clients[client].statements.length;

/**
 * Client correct statements include given statement
 * @param {string} client The client ID
 * @param {string} stmId The statement ID
 * @return {boolean}
 */
lib.clientHasStatement = (client, stmId) => clients[client].statements.includes(stmId);

/**
 * Determine statement correctness and lock state
 * @param {string} client The client ID
 * @param {boolean} correct Client answered correctly
 * @param {string} stmId The statement ID
 * @return {object}
 */
export const getStatementStatus = (client, correct, stmId) => {
  const status = {
    disabled: false,
    selected: false
  };

  if (correct) {
    status.disabled = true;
    status.selected = lib.clientHasStatement(client, stmId);
  }

  return status;
};

/**
 * Determine the lock duration in seconds
 * @param {number} attempts The number of attempts
 * @return {number}
 */
export const getClientLockedDuration = (attempts) => {
  const times = attempts / REQUIREMENT_MAX_ATTEMPTS;
  return CLIENT_LOCKED[times] || CLIENT_LOCKED.max;
};

/**
 * Get the remaining locked duration in seconds
 * @param {number} start The time the client was locked
 * @param {number} duration The lock duration (in seconds)
 * @return {number}
 */
export const getRemainingLockedDuration = (start, duration) => {
  let remaining = 0;
  const end = start + duration * 1000;
  const current = new Date().getTime();

  if (end > current) {
    remaining = Math.round((end - current) / 1000);
  }

  return remaining;
};

/**
 * Gets the remaining attempts
 * @param {number} attempts Current attempts
 * @return {number}
 */
export const getRemainingAttempts = attempts => (
  (attempts % REQUIREMENT_MAX_ATTEMPTS) === 0
    ? 0
    : REQUIREMENT_MAX_ATTEMPTS - (attempts % REQUIREMENT_MAX_ATTEMPTS)
);

/**
 * Returns a list of the requirement statements for a client
 * @param {string} client The client ID
 * @return {array}
 */
export const getRequirementStatements = client => (
  Object.values(I18n.t(`clientprops.clients.${client}.statements`))
);

/**
 * Returns a list of the requirement statement IDs
 * @param {string} client The client ID
 * @return {array}
 */
export const getRequirementStatementIds = (client) => {
  const data = I18n.t(`clientprops.clients.${client}.statements`);
  return Object.keys(data)
    .map(value => I18n.t(`clientprops.clients.${client}.statements.${value}.id`));
};

/**
 * Check to see if the selected statement is correct
 * @param {string} client The client ID
 * @param {string} solution The solution ID
 * @return {boolean}
 */
export const checkSolutionCorrectness = (client, solution) => (
  clients[client].solutions.includes(solution)
);

/**
 * Returns the number of correct solutions for a given client
 * @param {string} client The client ID
 * @return {number}
 */
export const getNumberOfCorrectSolutions = client => clients[client].solutions.length;

/**
 * Returns a list of the requirement statement IDs
 * @param {string} client The client ID
 * @return {array}
 */
export const getClientDifficultyScore = (client) => {
  const difficulty = clients[client].difficulty;
  return STEP1_SCORES[difficulty];
};

/**
 * Determine solution correctness and lock state
 * @param {string} client The client ID
 * @param {boolean} correct Client answered correctly
 * @param {string} solId The solution ID
 * @return {object}
 */
export const getSolutionStatus = (client, correct, givenUp, solId) => {
  const status = {
    disabled: false,
    selected: false
  };

  if (correct) {
    status.disabled = true;
    status.selected = lib.clientHasSolution(client, solId);
  }

  if (givenUp) {
    status.disabled = true;
    status.selected = false;
  }

  return status;
};

/**
 * Client correct solutions include given solution
 * @param {string} client The client ID
 * @param {string} stmId The solution ID
 * @return {boolean}
 */
lib.clientHasSolution = (client, solId) => clients[client].solutions.includes(solId);

/**
 * Send an xAPI statement for every resource
 * @param {string} clientId The client ID
 * @param {string} resourceId The resource ID
 * @param {string} title The resouce title
 * @param {string} description The resource description
 * @param {object} timeAvailable The time the popup became available
 * @param {number} questionIndex The question number
 */
export const sendResourceStatement = (
  clientId,
  resourceId,
  title,
  description,
  timeAvailable,
  questionIndex
) => {
  sendInteractionData({
    type: 'experienced',
    activity: {
      id: `${clientsPropsReferrer}/${clientId}/question-${questionIndex}/${resourceId}`,
      type: 'slide',
      title,
      description
    },
    response: {
      timeAvailable,
      timeResponse: new Date()
    },
    parent: `${clientsPropsReferrer}/${clientId}/question-${questionIndex}`
  });
};

/**
 * Returns the score of a client proposal
 * @param {string} client The client ID
 * @param {number} noOfIncorrect The number of incorrect attempts
 * @return {number}
 */
export const getClientProposalScore = (client, noOfIncorrect) => {
  const difficulty = clients[client].difficulty;
  const score = STEP2_SCORES[difficulty] * (0.9 ** noOfIncorrect);
  return Math.round(score);
};
