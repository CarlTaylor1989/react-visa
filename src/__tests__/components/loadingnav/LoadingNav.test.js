/* global toJson */
import React from 'react'; // eslint-disable-line no-unused-vars
import { LoadingNav, mapStateToProps } from '../../../components/loadingnav/LoadingNav';
import * as xapi from '../../../tracking/xapi/XApiAdapter';
import { hideLoader } from '../../../lib/app';

jest.mock('../../../components/generic/GoToScreen');
jest.mock('../../../tracking/xapi/XApiAdapter');
jest.mock('../../../lib/app');

let mockProps;
beforeEach(() => {
  jest.resetAllMocks();
  mockProps = {
    addPrompts: jest.fn(),
    dataReady: false,
    learnerID: '',
    learnerName: '',
    lessonLocation: '/',
    restoreAchievements: jest.fn(),
    setCategories: jest.fn(),
    setDiagnosticData: jest.fn(),
    setGenericData: jest.fn(),
    setRank: jest.fn(),
    setScreenReferrer: jest.fn(),
    setSettings: jest.fn(),
    setScores: jest.fn(),
    setSlideIndex: jest.fn(),
    setStreak: jest.fn(),
    suspendData: {},
    suspendDataParsed: jest.fn(),
    updateTimedAchievements: jest.fn(),
    scormInitialised: false,
    setClientProposalData: jest.fn(),
    setTutorialLaunched: jest.fn()
  };
});

test('renders the LoadingNav component in prod mode', () => {
  const wrapper = shallow(<LoadingNav {...mockProps} />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

describe('method componentDidMount', () => {
  test('renders the LoadingNav component in dev mode', () => {
    // setup
    const wrapper = shallow(<LoadingNav {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().findAndGoToScreen = jest.fn();
    global.process.env.PLATFORM = 'local';

    // function under test
    wrapper.instance().componentDidMount();

    // expectation
    expect(wrapper.instance().findAndGoToScreen).toHaveBeenCalled();
  });

  test('renders the LoadingNav component in production mode with scorm adapter initialised', () => {
    // setup
    mockProps.scormInitialised = true;
    const wrapper = shallow(<LoadingNav {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().onScormReady = jest.fn();
    global.process.env.PLATFORM = 'production';

    // function under test
    wrapper.instance().componentDidMount();

    // expectation
    expect(wrapper.instance().onScormReady).toHaveBeenCalled();
  });

  test('renders the LoadingNav component in production mode with scorm adapter not initialised', () => { // eslint-disable-line max-len
    // setup
    const wrapper = shallow(<LoadingNav {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().findAndGoToScreen = jest.fn();
    wrapper.instance().onScormReady = jest.fn();
    global.process.env.PLATFORM = 'production';

    // function under test
    wrapper.instance().componentDidMount();

    // expectation
    expect(wrapper.instance().findAndGoToScreen).not.toHaveBeenCalled();
    expect(wrapper.instance().onScormReady).not.toHaveBeenCalled();
  });
});

describe('LoadingNav component did update', () => {
  test('scorm is initialised', () => {
    // setup
    const wrapper = shallow(<LoadingNav {...mockProps} />);
    wrapper.instance().onScormReady = jest.fn();
    wrapper.setProps({ scormInitialised: true });

    // function under test
    wrapper.update();

    // expectations
    expect(wrapper.instance().onScormReady).toHaveBeenCalled();
  });

  test('hides loading screen and navigates to the splash page', () => {
    // setup
    const wrapper = shallow(<LoadingNav {...mockProps} />);
    wrapper.instance().goToSplashScreen = jest.fn();
    wrapper.setProps({ dataReady: true });

    // function under test
    wrapper.update();

    // expectations
    expect(mockProps.updateTimedAchievements).toHaveBeenCalled();
    expect(hideLoader).toHaveBeenCalled();
    expect(wrapper.instance().goToSplashScreen).toHaveBeenCalled();
  });

  test('loading screen remains visible', () => {
    // setup
    const wrapper = shallow(<LoadingNav {...mockProps} />);
    wrapper.instance().findAndGoToScreen = jest.fn();
    wrapper.setProps({ dataReady: false });

    // function under test
    wrapper.update();

    // expectations
    expect(hideLoader).not.toHaveBeenCalled();
    expect(wrapper.instance().findAndGoToScreen).not.toHaveBeenCalled();
  });
});

describe('method onScormReady', () => {
  test('xapi will be initalised', () => {
    // setup
    const scormData = {
      learnerID: 'aa@aa.aa',
      learnerName: 'irrelevant',
      lessonLocation: '/splash'
    };
    const wrapper = shallow(<LoadingNav {...mockProps} {...scormData} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().retrieveDataFromSuspendData = jest.fn();
    global.process.env.PLATFORM = 'production';

    // function under test
    wrapper.instance().onScormReady();

    // expectations
    expect(xapi.initialiseLrs).toHaveBeenCalledWith(
      expect.objectContaining({
        id: scormData.learnerID,
        name: scormData.learnerName
      }),
      true
    );
    expect(wrapper.instance().retrieveDataFromSuspendData).not.toHaveBeenCalled();
    expect(mockProps.suspendDataParsed).toHaveBeenCalledWith(true);
    expect(hideLoader).not.toHaveBeenCalled();
  });

  test('retrieves data from the suspend data', () => {
    // setup
    const scormData = {
      learnerID: 'aa@aa.aa',
      learnerName: 'irrelevant',
      lessonLocation: '/splash',
      suspendData: { foo: 'bar' }
    };
    const wrapper = shallow(<LoadingNav {...mockProps} {...scormData} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().retrieveDataFromSuspendData = jest.fn();
    global.process.env.PLATFORM = 'production';

    // function under test
    wrapper.instance().onScormReady();

    // expectations
    expect(wrapper.instance().retrieveDataFromSuspendData).toHaveBeenCalledWith(
      scormData.suspendData
    );
    expect(mockProps.suspendDataParsed).toHaveBeenCalledWith(true);
    expect(hideLoader).not.toHaveBeenCalled();
  });

  test('xapi cannot be initalised', () => {
    // setup
    const scormData = {
      learnerID: '',
      learnerName: '',
      lessonLocation: '/'
    };
    const wrapper = shallow(<LoadingNav {...mockProps} {...scormData} />);
    wrapper.instance().retrieveDataFromSuspendData = jest.fn();

    // function under test
    wrapper.instance().onScormReady();

    // expectations
    expect(xapi.initialiseLrs).not.toHaveBeenCalled();
    expect(wrapper.instance().retrieveDataFromSuspendData).not.toHaveBeenCalled();
    expect(hideLoader).not.toHaveBeenCalled();
  });
});

describe('method retrieveDataFromSuspendData()', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<LoadingNav {...mockProps} />);
  });

  test('empty suspend data', () => {
    wrapper.instance().retrieveDataFromSuspendData({});
    expect(mockProps.setSettings).not.toHaveBeenCalled();
  });

  test('suspend data with settings', () => {
    wrapper.instance().retrieveDataFromSuspendData({
      settings: {
        foo: 'bar'
      }
    });
    expect(mockProps.setSettings).toHaveBeenCalledWith(expect.any(Object));
  });

  describe('suspend data with diagnostic data', () => {
    test('diagnostic not completed', () => {
      wrapper.instance().retrieveDataFromSuspendData({
        diagnostic: false
      });
      expect(mockProps.setDiagnosticData).not.toHaveBeenCalled();
    });

    test('diagnostic completed', () => {
      wrapper.instance().retrieveDataFromSuspendData({
        diagnostic: true
      });
      expect(mockProps.setDiagnosticData).toHaveBeenCalledWith(true);
    });
  });

  describe('suspend data with scores', () => {
    test('scores with categories', () => {
      wrapper.instance().retrieveDataFromSuspendData({
        score: {
          rank: 'irrelevant',
          running: 1,
          total: 2,
          categories: { foo: 1 }
        }
      });
      expect(mockProps.setRank).toHaveBeenCalledWith(expect.any(String));
      expect(mockProps.setScores).toHaveBeenCalledWith(expect.any(Number), expect.any(Number));
      expect(mockProps.setCategories).toHaveBeenCalledWith(expect.any(Object));
    });

    test('scores without categories', () => {
      wrapper.instance().retrieveDataFromSuspendData({
        score: {
          rank: 'irrelevant',
          running: 1,
          total: 2
        }
      });
      expect(mockProps.setRank).toHaveBeenCalledWith(expect.any(String));
      expect(mockProps.setScores).toHaveBeenCalledWith(expect.any(Number), expect.any(Number));
      expect(mockProps.setCategories).not.toHaveBeenCalled();
    });
  });

  test('suspend data with streak', () => {
    wrapper.instance().retrieveDataFromSuspendData({
      streak: {
        active: false,
        completed: true
      }
    });
    expect(mockProps.setStreak).toHaveBeenCalledWith(expect.any(Object));
  });

  test('suspend data with generic data', () => {
    wrapper.instance().retrieveDataFromSuspendData({
      generic: {
        foo: false,
        bar: true
      }
    });
    expect(mockProps.setGenericData).toHaveBeenCalledWith(expect.any(Object));
  });

  test('suspend data with prompt data', () => {
    wrapper.instance().retrieveDataFromSuspendData({
      prompts: {
        foo: false,
        bar: true
      }
    });
    expect(mockProps.addPrompts).toHaveBeenCalledWith(expect.any(Object));
  });

  test('suspend data with achievements data', () => {
    const achievements = {
      completed: ['foo'],
      lastCompleted: 'foo'
    };
    wrapper.instance().retrieveDataFromSuspendData({ achievements });
    expect(mockProps.restoreAchievements).toHaveBeenCalledWith(
      achievements.completed,
      achievements.lastCompleted
    );
  });

  test('suspend data with clientproposal data', () => {
    wrapper.instance().retrieveDataFromSuspendData({
      clientProposalData: {
        foo: false,
        bar: true
      }
    });
    expect(mockProps.setClientProposalData).toHaveBeenCalledWith(expect.any(Object));
  });

  test('suspend data with tutorial data', () => {
    wrapper.instance().retrieveDataFromSuspendData({
      tutorial: {
        launched: true
      }
    });
    expect(mockProps.setTutorialLaunched).toHaveBeenCalledWith(expect.any(Boolean));
  });
});

test('should mapStateToProps', () => {
  const state = {
    genericData: {
      dataReady: false
    },
    scormData: {
      learnerID: '',
      learnerName: '',
      lessonLocation: '',
      suspendData: {}
    }
  };
  const props = mapStateToProps(state);
  expect(props).toEqual({
    ...state.genericData,
    ...state.scormData
  });
});
