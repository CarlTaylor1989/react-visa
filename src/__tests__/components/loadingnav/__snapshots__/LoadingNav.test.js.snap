// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the LoadingNav component in prod mode 1`] = `""`;
